File: gospider/captcha/solver.go
package captcha

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

// CaptchaSolver 验证码解决器
type CaptchaSolver struct {
	apiKey    string
	service   string
	timeout   time.Duration
	httpClient *http.Client
	pollInterval time.Duration
	maxPolls int
}

// NewCaptchaSolver 创建验证码解决器
func NewCaptchaSolver(apiKey, service string) *CaptchaSolver {
	return &CaptchaSolver{
		apiKey:  apiKey,
		service: service,
		timeout: 30 * time.Second,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
		pollInterval: 2 * time.Second,
		maxPolls: 30,
	}
}

// SolveResult 解决结果
type SolveResult struct {
	Success bool   `json:"success"`
	Text    string `json:"text"`
	Error   string `json:"error,omitempty"`
}

// SolveImage 解决图片验证码
func (cs *CaptchaSolver) SolveImage(imageData []byte) (*SolveResult, error) {
	switch cs.service {
	case "2captcha":
		return cs.solve2Captcha(imageData)
	case "anticaptcha":
		return cs.solveAntiCaptcha(imageData)
	default:
		return cs.solve2Captcha(imageData)
	}
}

// solve2Captcha 使用 2Captcha 解决
func (cs *CaptchaSolver) solve2Captcha(imageData []byte) (*SolveResult, error) {
	// 上传图片
	base64Image := base64.StdEncoding.EncodeToString(imageData)
	
	resp, err := cs.httpClient.PostForm(
		"http://2captcha.com/in.php",
		map[string][]string{
			"key":   {cs.apiKey},
			"method": {"base64"},
			"body":  {base64Image},
		},
	)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	
	body, _ := io.ReadAll(resp.Body)
	
	// 解析任务 ID
	var result map[string]string
	json.Unmarshal(body, &result)
	
	if result["status"] != "1" {
		return &SolveResult{
			Success: false,
			Error:   result["request"],
		}, nil
	}
	
	taskID := result["request"]
	
	// 轮询获取结果
	for i := 0; i < 30; i++ {
		time.Sleep(2 * time.Second)
		
		resp, err := cs.httpClient.Get(
			"http://2captcha.com/res.php?key=" + cs.apiKey + "&action=get&id=" + taskID,
		)
		if err != nil {
			continue
		}
		
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		
		var result map[string]string
		json.Unmarshal(body, &result)
		
		if result["status"] == "1" {
			return &SolveResult{
				Success: true,
				Text:    result["request"],
			}, nil
		}
		
		if result["request"] != "CAPCHA_NOT_READY" {
			return &SolveResult{
				Success: false,
				Error:   result["request"],
			}, nil
		}
	}
	
	return &SolveResult{
		Success: false,
		Error:   "Timeout",
	}, nil
}

// solveAntiCaptcha 使用 Anti-Captcha 解决
func (cs *CaptchaSolver) solveAntiCaptcha(imageData []byte) (*SolveResult, error) {
	// 实现 Anti-Captcha API 调用
	base64Image := base64.StdEncoding.EncodeToString(imageData)
	
	payload := map[string]interface{}{
		"clientKey": cs.apiKey,
		"task": map[string]interface{}{
			"type": "ImageToTextTask",
			"body": base64Image,
		},
	}
	
	jsonData, _ := json.Marshal(payload)
	
	resp, err := cs.httpClient.Post(
		"https://api.anti-captcha.com/createTask",
		"application/json",
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	
	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)
	
	if result["errorId"].(float64) != 0 {
		return &SolveResult{
			Success: false,
			Error:   result["errorDescription"].(string),
		}, nil
	}
	
	taskID := int(result["taskId"].(float64))
	
	// 轮询获取结果
	for i := 0; i < 30; i++ {
		time.Sleep(2 * time.Second)
		
		payload := map[string]interface{}{
			"clientKey": cs.apiKey,
			"taskId":    taskID,
		}
		
		jsonData, _ := json.Marshal(payload)
		
		resp, err := cs.httpClient.Post(
			"https://api.anti-captcha.com/getTaskResult",
			"application/json",
			bytes.NewBuffer(jsonData),
		)
		if err != nil {
			continue
		}
		
		var result map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&result)
		resp.Body.Close()
		
		if result["status"].(string) == "ready" {
			solution := result["solution"].(map[string]interface{})
			return &SolveResult{
				Success: true,
				Text:    solution["text"].(string),
			}, nil
		}
	}
	
	return &SolveResult{
		Success: false,
		Error:   "Timeout",
	}, nil
}

// SolveReCaptcha 解决 reCAPTCHA
func (cs *CaptchaSolver) SolveReCaptcha(siteKey, pageURL string) (*SolveResult, error) {
	if cs.apiKey == "" {
		return &SolveResult{Success: false, Error: "API key is required"}, nil
	}
	if siteKey == "" || pageURL == "" {
		return &SolveResult{Success: false, Error: "siteKey and pageURL are required"}, nil
	}

	switch cs.service {
	case "2captcha":
		token, err := cs.solve2CaptchaReCaptcha(siteKey, pageURL)
		if err != nil {
			return &SolveResult{Success: false, Error: err.Error()}, nil
		}
		return &SolveResult{Success: true, Text: token}, nil
	case "anticaptcha":
		token, err := cs.solveAntiCaptchaReCaptcha(siteKey, pageURL)
		if err != nil {
			return &SolveResult{Success: false, Error: err.Error()}, nil
		}
		return &SolveResult{Success: true, Text: token}, nil
	default:
		return &SolveResult{Success: false, Error: "unsupported service"}, nil
	}
}

// SolveHCaptcha 解决 hCaptcha
func (cs *CaptchaSolver) SolveHCaptcha(siteKey, pageURL string) (*SolveResult, error) {
	if cs.apiKey == "" {
		return &SolveResult{Success: false, Error: "API key is required"}, nil
	}
	if siteKey == "" || pageURL == "" {
		return &SolveResult{Success: false, Error: "siteKey and pageURL are required"}, nil
	}

	switch cs.service {
	case "2captcha":
		token, err := cs.solve2CaptchaHCaptcha(siteKey, pageURL)
		if err != nil {
			return &SolveResult{Success: false, Error: err.Error()}, nil
		}
		return &SolveResult{Success: true, Text: token}, nil
	case "anticaptcha":
		token, err := cs.solveAntiCaptchaHCaptcha(siteKey, pageURL)
		if err != nil {
			return &SolveResult{Success: false, Error: err.Error()}, nil
		}
		return &SolveResult{Success: true, Text: token}, nil
	default:
		return &SolveResult{Success: false, Error: "unsupported service"}, nil
	}
}

// ReportBad 报告错误的识别结果
func (cs *CaptchaSolver) ReportBad(taskID string) error {
	// 实现报告错误识别
	return nil
}

// SetPollingConfig 调整轮询间隔与最大轮询次数，便于测试和慢速 provider 调优。
func (cs *CaptchaSolver) SetPollingConfig(interval time.Duration, maxPolls int) {
	if interval > 0 {
		cs.pollInterval = interval
	}
	if maxPolls > 0 {
		cs.maxPolls = maxPolls
	}
}

func (cs *CaptchaSolver) solve2CaptchaReCaptcha(siteKey, pageURL string) (string, error) {
	formData := url.Values{}
	formData.Set("key", cs.apiKey)
	formData.Set("method", "userrecaptcha")
	formData.Set("googlekey", siteKey)
	formData.Set("pageurl", pageURL)
	formData.Set("json", "1")

	taskID, err := cs.submit2CaptchaTask(formData)
	if err != nil {
		return "", err
	}
	return cs.poll2CaptchaTask(taskID)
}

func (cs *CaptchaSolver) solve2CaptchaHCaptcha(siteKey, pageURL string) (string, error) {
	formData := url.Values{}
	formData.Set("key", cs.apiKey)
	formData.Set("method", "hcaptcha")
	formData.Set("sitekey", siteKey)
	formData.Set("pageurl", pageURL)
	formData.Set("json", "1")

	taskID, err := cs.submit2CaptchaTask(formData)
	if err != nil {
		return "", err
	}
	return cs.poll2CaptchaTask(taskID)
}

func (cs *CaptchaSolver) solveAntiCaptchaReCaptcha(siteKey, pageURL string) (string, error) {
	taskID, err := cs.createAntiCaptchaTask(map[string]interface{}{
		"type":       "NoCaptchaTaskProxyless",
		"websiteURL": pageURL,
		"websiteKey": siteKey,
	})
	if err != nil {
		return "", err
	}
	return cs.pollAntiCaptchaTask(taskID, "gRecaptchaResponse")
}

func (cs *CaptchaSolver) solveAntiCaptchaHCaptcha(siteKey, pageURL string) (string, error) {
	taskID, err := cs.createAntiCaptchaTask(map[string]interface{}{
		"type":       "HCaptchaTaskProxyless",
		"websiteURL": pageURL,
		"websiteKey": siteKey,
	})
	if err != nil {
		return "", err
	}
	return cs.pollAntiCaptchaTask(taskID, "gRecaptchaResponse")
}

func (cs *CaptchaSolver) submit2CaptchaTask(formData url.Values) (string, error) {
	resp, err := cs.httpClient.PostForm("https://2captcha.com/in.php", formData)
	if err != nil {
		return "", fmt.Errorf("submit 2captcha task: %w", err)
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("decode 2captcha response: %w", err)
	}
	if status, ok := result["status"].(float64); !ok || status != 1 {
		return "", fmt.Errorf("2captcha submit failed: %v", result["request"])
	}
	taskID, ok := result["request"].(string)
	if !ok || taskID == "" {
		return "", fmt.Errorf("2captcha returned invalid task id")
	}
	return taskID, nil
}

func (cs *CaptchaSolver) poll2CaptchaTask(taskID string) (string, error) {
	for i := 0; i < cs.maxPolls; i++ {
		time.Sleep(cs.pollInterval)
		resp, err := cs.httpClient.Get(
			"https://2captcha.com/res.php?key=" + cs.apiKey + "&action=get&id=" + taskID + "&json=1",
		)
		if err != nil {
			continue
		}

		var result map[string]interface{}
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			resp.Body.Close()
			continue
		}
		resp.Body.Close()

		if status, ok := result["status"].(float64); ok && status == 1 {
			if request, ok := result["request"].(string); ok && request != "" {
				return request, nil
			}
		}
		if request, ok := result["request"].(string); ok && request != "CAPCHA_NOT_READY" {
			return "", fmt.Errorf("2captcha poll failed: %s", request)
		}
	}
	return "", fmt.Errorf("2captcha solve timeout")
}

func (cs *CaptchaSolver) createAntiCaptchaTask(task map[string]interface{}) (int, error) {
	payload := map[string]interface{}{
		"clientKey": cs.apiKey,
		"task":      task,
	}
	jsonData, _ := json.Marshal(payload)
	resp, err := cs.httpClient.Post(
		"https://api.anti-captcha.com/createTask",
		"application/json",
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		return 0, fmt.Errorf("submit anti-captcha task: %w", err)
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return 0, fmt.Errorf("decode anti-captcha response: %w", err)
	}
	if errorID, ok := result["errorId"].(float64); ok && errorID != 0 {
		return 0, fmt.Errorf("anti-captcha submit failed: %v", result["errorDescription"])
	}
	taskID, ok := result["taskId"].(float64)
	if !ok {
		return 0, fmt.Errorf("anti-captcha returned invalid task id")
	}
	return int(taskID), nil
}

func (cs *CaptchaSolver) pollAntiCaptchaTask(taskID int, field string) (string, error) {
	for i := 0; i < cs.maxPolls; i++ {
		time.Sleep(cs.pollInterval)
		payload := map[string]interface{}{
			"clientKey": cs.apiKey,
			"taskId":    taskID,
		}
		jsonData, _ := json.Marshal(payload)
		resp, err := cs.httpClient.Post(
			"https://api.anti-captcha.com/getTaskResult",
			"application/json",
			bytes.NewBuffer(jsonData),
		)
		if err != nil {
			continue
		}

		var result map[string]interface{}
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			resp.Body.Close()
			continue
		}
		resp.Body.Close()

		if errorID, ok := result["errorId"].(float64); ok && errorID != 0 {
			return "", fmt.Errorf("anti-captcha poll failed: %v", result["errorDescription"])
		}
		if status, ok := result["status"].(string); ok && status == "ready" {
			if solution, ok := result["solution"].(map[string]interface{}); ok {
				if token, ok := solution[field].(string); ok && token != "" {
					return token, nil
				}
			}
			return "", fmt.Errorf("anti-captcha missing solution token")
		}
	}
	return "", fmt.Errorf("anti-captcha solve timeout")
}

File: rustspider/src/captcha/solver.rs
//! 验证码识别模块

use std::collections::BTreeMap;
use std::time::Duration;

/// 验证码解决器
pub struct CaptchaSolver {
    api_key: String,
    service: String,
    timeout: Duration,
    client: reqwest::blocking::Client,
    poll_interval: Duration,
    max_polls: usize,
    base_url: Option<String>,
}

/// 解决结果
#[derive(Debug, Clone)]
pub struct SolveResult {
    pub success: bool,
    pub text: Option<String>,
    pub error: Option<String>,
}

impl CaptchaSolver {
    /// 创建验证码解决器
    pub fn new(api_key: &str, service: &str) -> Self {
        CaptchaSolver {
            api_key: api_key.to_string(),
            service: service.to_string(),
            timeout: Duration::from_secs(30),
            client: reqwest::blocking::Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .unwrap_or_default(),
            poll_interval: Duration::from_secs(2),
            max_polls: 30,
            base_url: None,
        }
    }

    pub fn set_polling_config(&mut self, poll_interval: Duration, max_polls: usize) {
        if poll_interval > Duration::ZERO {
            self.poll_interval = poll_interval;
        }
        if max_polls > 0 {
            self.max_polls = max_polls;
        }
    }

    pub fn set_base_url(&mut self, base_url: &str) {
        self.base_url = Some(base_url.trim_end_matches('/').to_string());
    }

    /// 解决图片验证码
    pub fn solve_image(&self, image_data: &[u8]) -> SolveResult {
        match self.service.as_str() {
            "2captcha" => self.solve_2captcha(image_data),
            "anticaptcha" => self.solve_anticaptcha(image_data),
            _ => self.solve_2captcha(image_data),
        }
    }

    /// 使用 2Captcha 解决
    fn solve_2captcha(&self, image_data: &[u8]) -> SolveResult {
        // 上传图片
        let base64_image = encode_base64(image_data);

        let mut payload = BTreeMap::new();
        payload.insert("key".to_string(), self.api_key.clone());
        payload.insert("method".to_string(), "base64".to_string());
        payload.insert("body".to_string(), base64_image);

        let resp = match self
            .client
            .post(&self.build_url("http://2captcha.com/in.php"))
            .form(&payload)
            .send()
        {
            Ok(r) => r,
            Err(e) => {
                return SolveResult {
                    success: false,
                    text: None,
                    error: Some(e.to_string()),
                }
            }
        };

        let text = resp.text().unwrap_or_default();
        let parts: Vec<&str> = text.split('|').collect();

        if parts[0] != "OK" {
            return SolveResult {
                success: false,
                text: None,
                error: Some(parts[0].to_string()),
            };
        }

        let task_id = parts[1];

        // 轮询获取结果
        for _ in 0..30 {
            std::thread::sleep(Duration::from_secs(2));

            let resp = match self
                .client
                .get(&format!(
                    "{}?key={}&action=get&id={}",
                    self.build_url("http://2captcha.com/res.php"),
                    self.api_key,
                    task_id
                ))
                .send()
            {
                Ok(r) => r,
                Err(e) => continue,
            };

            let text = resp.text().unwrap_or_default();
            let parts: Vec<&str> = text.split('|').collect();

            if parts[0] == "OK" {
                return SolveResult {
                    success: true,
                    text: Some(parts[1].to_string()),
                    error: None,
                };
            }

            if parts[0] != "CAPCHA_NOT_READY" {
                return SolveResult {
                    success: false,
                    text: None,
                    error: Some(parts[0].to_string()),
                };
            }
        }

        SolveResult {
            success: false,
            text: None,
            error: Some("Timeout".to_string()),
        }
    }

    /// 使用 Anti-Captcha 解决
    fn solve_anticaptcha(&self, image_data: &[u8]) -> SolveResult {
        let base64_image = encode_base64(image_data);

        // 创建任务
        let resp = match self
            .client
            .post(&self.build_url("https://api.anti-captcha.com/createTask"))
            .json(&serde_json::json!({
                "clientKey": self.api_key,
                "task": {
                    "type": "ImageToTextTask",
                    "body": base64_image,
                },
            }))
            .send()
        {
            Ok(r) => r,
            Err(e) => {
                return SolveResult {
                    success: false,
                    text: None,
                    error: Some(e.to_string()),
                }
            }
        };

        let result: serde_json::Value = resp.json().unwrap_or_default();

        if result["errorId"].as_i64().unwrap_or(1) != 0 {
            return SolveResult {
                success: false,
                text: None,
                error: Some(
                    result["errorDescription"]
                        .as_str()
                        .unwrap_or("Unknown error")
                        .to_string(),
                ),
            };
        }

        let task_id = result["taskId"].as_i64().unwrap_or(0);

        // 轮询获取结果
        for _ in 0..30 {
            std::thread::sleep(Duration::from_secs(2));

            let resp = match self
                .client
                .post(&self.build_url("https://api.anti-captcha.com/getTaskResult"))
                .json(&serde_json::json!({
                    "clientKey": self.api_key,
                    "taskId": task_id,
                }))
                .send()
            {
                Ok(r) => r,
                Err(e) => continue,
            };

            let result: serde_json::Value = resp.json().unwrap_or_default();

            if result["status"].as_str() == Some("ready") {
                if let Some(text) = result["solution"]["text"].as_str() {
                    return SolveResult {
                        success: true,
                        text: Some(text.to_string()),
                        error: None,
                    };
                }
            }
        }

        SolveResult {
            success: false,
            text: None,
            error: Some("Timeout".to_string()),
        }
    }

    /// 解决 reCAPTCHA
    pub fn solve_recaptcha(&self, site_key: &str, page_url: &str) -> SolveResult {
        if self.api_key.is_empty() {
            return SolveResult {
                success: false,
                text: None,
                error: Some("API key is required".to_string()),
            };
        }

        match self.service.as_str() {
            "2captcha" => {
                match self.solve_2captcha_token("userrecaptcha", site_key, page_url, "googlekey") {
                    Ok(token) => SolveResult {
                        success: true,
                        text: Some(token),
                        error: None,
                    },
                    Err(err) => SolveResult {
                        success: false,
                        text: None,
                        error: Some(err.to_string()),
                    },
                }
            }
            "anticaptcha" => {
                match self.solve_anticaptcha_token("NoCaptchaTaskProxyless", site_key, page_url) {
                    Ok(token) => SolveResult {
                        success: true,
                        text: Some(token),
                        error: None,
                    },
                    Err(err) => SolveResult {
                        success: false,
                        text: None,
                        error: Some(err.to_string()),
                    },
                }
            }
            _ => SolveResult {
                success: false,
                text: None,
                error: Some("unsupported service".to_string()),
            },
        }
    }

    /// 解决 hCaptcha
    pub fn solve_hcaptcha(&self, site_key: &str, page_url: &str) -> SolveResult {
        if self.api_key.is_empty() {
            return SolveResult {
                success: false,
                text: None,
                error: Some("API key is required".to_string()),
            };
        }

        match self.service.as_str() {
            "2captcha" => {
                match self.solve_2captcha_token("hcaptcha", site_key, page_url, "sitekey") {
                    Ok(token) => SolveResult {
                        success: true,
                        text: Some(token),
                        error: None,
                    },
                    Err(err) => SolveResult {
                        success: false,
                        text: None,
                        error: Some(err.to_string()),
                    },
                }
            }
            "anticaptcha" => {
                match self.solve_anticaptcha_token("HCaptchaTaskProxyless", site_key, page_url) {
                    Ok(token) => SolveResult {
                        success: true,
                        text: Some(token),
                        error: None,
                    },
                    Err(err) => SolveResult {
                        success: false,
                        text: None,
                        error: Some(err.to_string()),
                    },
                }
            }
            _ => SolveResult {
                success: false,
                text: None,
                error: Some("unsupported service".to_string()),
            },
        }
    }

    /// 报告错误的识别结果
    pub fn report_bad(&self, task_id: &str) -> Result<(), Box<dyn std::error::Error>> {
        // 实现报告错误识别
        Ok(())
    }

    fn solve_2captcha_token(
        &self,
        method: &str,
        site_key: &str,
        page_url: &str,
        site_key_field: &str,
    ) -> Result<String, Box<dyn std::error::Error>> {
        let submit_url = self.build_url("https://2captcha.com/in.php");
        let task_id = self
            .client
            .post(&submit_url)
            .form(&[
                ("key", self.api_key.as_str()),
                ("method", method),
                (site_key_field, site_key),
                ("pageurl", page_url),
                ("json", "1"),
            ])
            .send()?
            .json::<serde_json::Value>()?;

        if task_id["status"].as_i64().unwrap_or_default() != 1 {
            return Err(format!(
                "2captcha submit failed: {}",
                task_id["request"].as_str().unwrap_or("unknown")
            )
            .into());
        }

        let task_id = task_id["request"]
            .as_str()
            .ok_or("2captcha returned invalid task id")?;
        self.poll_2captcha_token(task_id)
    }

    fn poll_2captcha_token(&self, task_id: &str) -> Result<String, Box<dyn std::error::Error>> {
        for _ in 0..self.max_polls {
            std::thread::sleep(self.poll_interval);
            let poll_url = format!(
                "{}?key={}&action=get&id={}&json=1",
                self.build_url("https://2captcha.com/res.php"),
                self.api_key,
                task_id
            );
            let result = self
                .client
                .get(&poll_url)
                .send()?
                .json::<serde_json::Value>()?;
            if result["status"].as_i64().unwrap_or_default() == 1 {
                return Ok(result["request"].as_str().unwrap_or_default().to_string());
            }
            let request = result["request"].as_str().unwrap_or_default();
            if request != "CAPCHA_NOT_READY" {
                return Err(format!("2captcha poll failed: {request}").into());
            }
        }
        Err("2captcha solve timeout".into())
    }

    fn solve_anticaptcha_token(
        &self,
        task_type: &str,
        site_key: &str,
        page_url: &str,
    ) -> Result<String, Box<dyn std::error::Error>> {
        let create_url = self.build_url("https://api.anti-captcha.com/createTask");
        let result = self
            .client
            .post(&create_url)
            .json(&serde_json::json!({
                "clientKey": self.api_key,
                "task": {
                    "type": task_type,
                    "websiteURL": page_url,
                    "websiteKey": site_key
                }
            }))
            .send()?
            .json::<serde_json::Value>()?;

        if result["errorId"].as_i64().unwrap_or(1) != 0 {
            return Err(result["errorDescription"]
                .as_str()
                .unwrap_or("anti-captcha create task failed")
                .to_string()
                .into());
        }
        let task_id = result["taskId"]
            .as_i64()
            .ok_or("anti-captcha returned invalid task id")?;
        self.poll_anticaptcha_token(task_id)
    }

    fn poll_anticaptcha_token(&self, task_id: i64) -> Result<String, Box<dyn std::error::Error>> {
        let get_url = self.build_url("https://api.anti-captcha.com/getTaskResult");
        for _ in 0..self.max_polls {
            std::thread::sleep(self.poll_interval);
            let result = self
                .client
                .post(&get_url)
                .json(&serde_json::json!({
                    "clientKey": self.api_key,
                    "taskId": task_id
                }))
                .send()?
                .json::<serde_json::Value>()?;

            if result["errorId"].as_i64().unwrap_or(1) != 0 {
                return Err(result["errorDescription"]
                    .as_str()
                    .unwrap_or("anti-captcha poll failed")
                    .to_string()
                    .into());
            }
            if result["status"].as_str() == Some("ready") {
                let solution = &result["solution"];
                let token = solution["gRecaptchaResponse"]
                    .as_str()
                    .or_else(|| solution["token"].as_str())
                    .unwrap_or_default();
                if !token.is_empty() {
                    return Ok(token.to_string());
                }
            }
        }
        Err("anti-captcha solve timeout".into())
    }

    fn build_url(&self, default_url: &str) -> String {
        if let Some(base) = &self.base_url {
            if default_url.contains("/in.php") {
                return format!("{base}/in.php");
            }
            if default_url.contains("/res.php") {
                return format!("{base}/res.php");
            }
            if default_url.contains("/createTask") {
                return format!("{base}/createTask");
            }
            if default_url.contains("/getTaskResult") {
                return format!("{base}/getTaskResult");
            }
        }
        default_url.to_string()
    }
}

fn encode_base64(input: &[u8]) -> String {
    const TABLE: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    let mut output = String::with_capacity(input.len().div_ceil(3) * 4);
    let mut chunks = input.chunks_exact(3);
    for chunk in &mut chunks {
        let value = ((chunk[0] as u32) << 16) | ((chunk[1] as u32) << 8) | chunk[2] as u32;
        output.push(TABLE[((value >> 18) & 0x3f) as usize] as char);
        output.push(TABLE[((value >> 12) & 0x3f) as usize] as char);
        output.push(TABLE[((value >> 6) & 0x3f) as usize] as char);
        output.push(TABLE[(value & 0x3f) as usize] as char);
    }

    let remainder = chunks.remainder();
    if !remainder.is_empty() {
        let first = remainder[0] as u32;
        output.push(TABLE[((first >> 2) & 0x3f) as usize] as char);
        if remainder.len() == 1 {
            output.push(TABLE[((first & 0x03) << 4) as usize] as char);
            output.push('=');
            output.push('=');
        } else {
            let second = remainder[1] as u32;
            output.push(TABLE[(((first & 0x03) << 4) | ((second >> 4) & 0x0f)) as usize] as char);
            output.push(TABLE[((second & 0x0f) << 2) as usize] as char);
            output.push('=');
        }
    }

    output
}

/// Cloudflare 绕过器
pub struct CloudflareBypass {
    client: reqwest::blocking::Client,
}

impl CloudflareBypass {
    pub fn new() -> Self {
        CloudflareBypass {
            client: reqwest::blocking::Client::builder()
                .user_agent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")
                .build()
                .unwrap_or_default(),
        }
    }

    pub fn get_token(&self, url: &str) -> Option<String> {
        // 实现 Cloudflare 绕过逻辑
        None
    }

    pub fn solve_challenge(&self, html: &str) -> Option<String> {
        // 解析 JavaScript 挑战
        None
    }
}

impl Default for CloudflareBypass {
    fn default() -> Self {
        Self::new()
    }
}

/// Akamai 绕过器
pub struct AkamaiBypass {
    client: reqwest::blocking::Client,
}

impl AkamaiBypass {
    pub fn new() -> Self {
        AkamaiBypass {
            client: reqwest::blocking::Client::builder()
                .build()
                .unwrap_or_default(),
        }
    }

    pub fn get_sensor_data(&self, url: &str) -> Option<std::collections::HashMap<String, String>> {
        // 实现 Akamai 绕过逻辑
        None
    }
}

impl Default for AkamaiBypass {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::{Read, Write};
    use std::net::TcpListener;
    use std::sync::atomic::{AtomicBool, Ordering};
    use std::sync::Arc;
    use std::thread;

    struct MockServer {
        addr: String,
        shutdown: Arc<AtomicBool>,
        handle: Option<thread::JoinHandle<()>>,
    }

    impl Drop for MockServer {
        fn drop(&mut self) {
            self.shutdown.store(true, Ordering::SeqCst);
            let _ = std::net::TcpStream::connect(&self.addr);
            if let Some(handle) = self.handle.take() {
                let _ = handle.join();
            }
        }
    }

    fn start_mock_server(handler: fn(&str) -> String) -> MockServer {
        let listener = TcpListener::bind("127.0.0.1:0").expect("listener");
        listener.set_nonblocking(true).expect("nonblocking");
        let addr = listener.local_addr().expect("addr");
        let shutdown = Arc::new(AtomicBool::new(false));
        let flag = shutdown.clone();
        let handle = thread::spawn(move || {
            while !flag.load(Ordering::SeqCst) {
                match listener.accept() {
                    Ok((mut stream, _)) => {
                        let mut buffer = [0_u8; 8192];
                        let size = stream.read(&mut buffer).unwrap_or(0);
                        let request = String::from_utf8_lossy(&buffer[..size]).to_string();
                        let body = handler(&request);
                        let response = format!(
                            "HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}",
                            body.len(),
                            body
                        );
                        let _ = stream.write_all(response.as_bytes());
                        let _ = stream.flush();
                    }
                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
                        thread::sleep(Duration::from_millis(20));
                    }
                    Err(_) => break,
                }
            }
        });
        MockServer {
            addr: addr.to_string(),
            shutdown,
            handle: Some(handle),
        }
    }

    #[test]
    fn test_captcha_solver() {
        let solver = CaptchaSolver::new("test-key", "2captcha");
        // 实际测试需要有效的 API key 和图片数据
    }

    #[test]
    fn solve_recaptcha_with_2captcha() {
        let server = start_mock_server(|request| {
            if request.starts_with("POST /in.php") {
                r#"{"status":1,"request":"task-1"}"#.to_string()
            } else {
                r#"{"status":1,"request":"recaptcha-token"}"#.to_string()
            }
        });
        let mut solver = CaptchaSolver::new("demo-key", "2captcha");
        solver.set_base_url(&format!("http://{}", server.addr));
        solver.set_polling_config(Duration::from_millis(1), 2);

        let result = solver.solve_recaptcha("site-key", "https://example.com");
        assert!(result.success);
        assert_eq!(result.text.as_deref(), Some("recaptcha-token"));
    }

    #[test]
    fn solve_hcaptcha_with_anticaptcha() {
        let server = start_mock_server(|request| {
            if request.starts_with("POST /createTask") {
                r#"{"errorId":0,"taskId":42}"#.to_string()
            } else {
                r#"{"errorId":0,"status":"ready","solution":{"gRecaptchaResponse":"hcaptcha-token"}}"#.to_string()
            }
        });
        let mut solver = CaptchaSolver::new("demo-key", "anticaptcha");
        solver.set_base_url(&format!("http://{}", server.addr));
        solver.set_polling_config(Duration::from_millis(1), 2);

        let result = solver.solve_hcaptcha("site-key", "https://example.com");
        assert!(result.success);
        assert_eq!(result.text.as_deref(), Some("hcaptcha-token"));
    }
}
